File: java_code/p322_medium_coin_change.java
// class Solution
class coinChangeSolution {
    public int coinChange(int[] coins, int amount) {
        if (amount == 0) return 0;
        if (amount < 0 || coins.length == 0) return -1;
        // 0 for not set, -1 for impossible
        int[] result = new int[amount + 1];
        // init coins
        // don't  forget to discard coin larger than amount
        for (int coinValue : coins) {
            if (coinValue <= amount) result[coinValue] = 1;
        }
        for (int i = 1; i < result.length; i++) {
            if (result[i] != 0) continue;
            for (int coinValue : coins) {
                if (i < coinValue || result[i - coinValue] == -1) continue;
                int num = result[i - coinValue] + 1;
                if (result[i] > num || result[i] == 0) result[i] = num;
            }
            if (result[i] == 0) result[i] = -1;
        }

        return result[amount];
    }
}

File: python_code/p322_medium_coin_change.py
# assume all(coin>0 in coins)
# no use for coinValue>amount
# dummyHead: make index==coninValue
from math import ceil


class Solution:
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        # handle amount = 0 and < 0
        # O(n^2) n is amount, too slow when amount is huge
        if amount < 0:
            return -1
        if amount == 0:
            return 0

        result = [1 if x in coins else None for x in range(amount + 1)]
        result[0] = -1
        for i, v in enumerate(result):
            if v:
                continue
            for ii in range(ceil(i / 2) + 1):  # find smallest combination
                if result[ii] == -1 or result[i - ii] == -1:
                    continue
                num = result[ii] + result[i - ii]
                if result[i] is None or num < result[i]:
                    result[i] = num
            if result[i] is None:
                result[i] = -1
        return result[-1]


if __name__ == '__main__':
    xx = Solution()
    k = xx.coinChange([3, 7, 405, 436], 8839)
    print(k)  # too slow!
